### ISSUE TITLE
std.json.JSONException@std/json.d(442): Key not found: cTag

### ISSUE THREAD

**Issue by jdevney** [skilion/onedrive#116]
Evening,

I get the following exception after an initial sync. It syncs a good proportion of my files, but then bombs-out. Same error each time.

Output:
## std.json.JSONException@std/json.d(442): Key not found: cTag

??:? pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, const(char[])) [0x62a766]
??:? pure @safe inout(std.json.JSONValue)\* std.exception.enforce!(std.json.JSONException, inout(std.json.JSONValue)_).enforce(inout(std.json.JSONValue)_, lazy const(char)[], immutable(char)[], ulong) [0x62a7ca]
??:? std.json.JSONValue.opIndexinout(pure ref inout(std.json.JSONValue) function(immutable(char)[])) [0x5fc236]
??:? void sync.SyncEngine.applyDifference(std.json.JSONValue) [0x5d4736]
??:? void sync.SyncEngine.applyDifferences() [0x5d3e38]
??:? void main.performSync(sync.SyncEngine) [0x5c4cd3]
??:? _Dmain [0x5c42e4]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5dee66]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5dedb0]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])_).runAll() [0x5dee22]
??:? void rt.dmain2._d_run_main(int, char__, extern (C) int function(char[][])_).tryExec(scope void delegate()) [0x5dedb0]
??:? _d_run_main [0x5ded21]
??:? main [0x5c68e5]
??:? __libc_start_main [0x46afe82f]
I can run with a debug build if that helps?
Cheers, JD.

**Comment by jdevney** [comment]
So I've another machine on which I've installed onedrive & this error doesn't come up.

Any thoughts?

**Comment by a4z** [comment]
had exactly the same problem,  run onedrive the first time, it starts to sync, than it wants to upload something (even if no changes have been made!), than a crash
``` 
Uploading: ./Pictures/Camera Roll/WP_20140223_001.jpg
std.json.JSONException@std/json.d(473): Key not found: cTag
----------------
??:? pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, const(char[])) [0x62e469]
??:? pure @safe inout(std.json.JSONValue)* std.exception.enforce!(std.json.JSONException, inout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*, lazy const(char)[], immutable(char)[], ulong) [0x62e4ce]
??:? std.json.JSONValue.opIndexinout(pure ref @safe inout(std.json.JSONValue) function(immutable(char)[])) [0x5ee523]
??:? void sync.SyncEngine.saveItem(std.json.JSONValue) [0x5c41fa]
??:? void sync.SyncEngine.uploadFileDifferences(itemdb.Item, immutable(char)[]) [0x5c1d4f]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x5c1365]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x5c117c]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x5c117c]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x5c117c]
??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0x5c0b99]
??:? void main.performSync(sync.SyncEngine) [0x5af316]
??:? _Dmain [0x5aeb32]
??:? _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x5d546e]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5d53b8]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x5d542a]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x5d53b8]
??:? _d_run_main [0x5d5329]
??:? main [0x5b1bb5]
??:? __libc_start_main [0xb33aa82f]
Segmentation fault (core dumped)
```

**Comment by fsouza** [comment]
I'm seeing this failure everywhere I try to 

The interesting thing is that I'm just syncing a remote folder, not uploading anything. Looking at the code, looks like [this piece](https://github.com/skilion/onedrive/blob/eb8d0fe039290da1b536654995d02a7e272bb9d7/src/sync.d#L181-L188) is ready to handle the absence of cTag, but [this](https://github.com/skilion/onedrive/blob/eb8d0fe039290da1b536654995d02a7e272bb9d7/src/sync.d#L530) is not (the one that is failing).

Should all accesses to the cTag field in json items be protected by ``try { ... } catch (JSONException e) { ... }``? That would probably involve creating a helper method to avoid repetition (something like ``private string getValue(JSONValue item, string key)``, that would return an empty string if ``key`` is not present in ``item``).

**Comment by a4z** [comment]
the other half question is why does this tool want to upload a file straight after it downloaded all files, even if there has been no change made.
running again, with verbose option, onedrive detects a lot of changed files with last access time changed. I think this is not good, there have been made no changes to the files. not sure if I am brave enough to run this tool again, start to become afraid about my data.

**Comment by skilion** [comment]
Fixed in 9ad9394b9827ce3e2bd7ab290451b7959cfbad50
Also c9016845c339104e54df95efdca617afa9786ee7 may have fixed the last problem @a4z referenced
